### DISCUSSION TITLE
Support lazily loading messages

### DISCUSSION THREAD

**Discussion by marcaddeo** [maxcountryman/axum-messages#16] [Ideas]
I have a middleware that uses the `Messages` extractor, but if a condition is not met the messages are never actually used by the middleware. Since the extractor is calling `Message::load`, the messages are removed by my middleware and never able to be used by further handlers/middleware.

It would be nice if there were a second extractor that did not automatically load the messages, leaving it up to the handler/middleware to do so manually. For that to work, the `Messages::load` method would need to be made public.

Is this something you'd consider adding or merging if I were to put together a PR?

**Comment by maxcountryman** [comment]
I think it would be nice to support this as a configuration option if possible--a second extractor could also work but it would be a shame to introduce a separate extractor just to allow manual loading.

**Reply by marcaddeo** [reply]
For my use case, I’d be using both the normal automatic extractor and the lazy extractor. I use the normal Messages extractor in my handlers, but I also need to access the unloaded messages in my response middleware.

**Reply by marcaddeo** [reply]
Another thought is removing the messages from the message store at a later stage in the lifecycle, i.e. when it’s turned into an iterator instead of in `Messages::load`. I’m not sure how feasible that solution would be, a lazy extractor was my first attempt at solving this without changing too much code in `axum-messages`.

**Comment by maxcountryman** [comment]
> I have a middleware that uses the Messages extractor, but if a condition is not met the messages are never actually used by the middleware.

Couldn't you also work around this by loading the `Messages` extractor only if your condition is met? Specifically I mean by taking the request (as an extractor) and loading from its extensions manually.

It might even be possible to wrap this all in an extractor of your own that wraps `Messages` and the conditional logic. That would move all this logic out of your handler.

All this is say I think a better approach is to leverage Axum itself a little more as opposed to change this crate.

**Reply by marcaddeo** [reply]
Hmm, I believe I had tried this but ran into an issue. I'll give it another shot today!

**Reply by marcaddeo** [reply]
Oh right, I think because my middleware is a response middleware I can't use the `Request` extractor. Possibly because it uses `FromRequest`? The error is very unclear 😅 

However, if `Messages::load` were public implementing the lazy message extractor in my own crate would work great I believe.

**Reply by maxcountryman** [reply]
> because my middleware is a response middleware I can't use the Request extractor

Can you share a concrete example of what you're trying to do?

Note that `Messages` uses the request (via `FromRequestParts`, but `FromRequest` provides similar access) and so you should also be able to do the same. You may need to implement your own extractor (although I don't think that's necessarily required, but likely a cleaner solution overall).

**Reply by marcaddeo** [reply]
I'm using [`map_response_with_state`](https://docs.rs/axum/latest/axum/middleware/fn.map_response_with_state.html) which doesn't support extractors that use `FromRequest`, only ones that use `FromRequestParts`. I suspect the `Request` extractor is a `FromRequest`, but I can't seem to actually confirm it from the docs..

I also just realized I can use the `Extension<Messages>` extractor to get the unloaded messages, so shouldn't need a custom extractor at all.

However, either way once I get the unloaded messages I can't load them due to `Messages::load` being a private method.

> Can you share a concrete example of what you're trying to do?

Does that help explain enough? I can link to my repository I'm working on, but obviously it's got more code than just what's needed to show what I'm trying to do 😅

**Reply by maxcountryman** [reply]
You don't need to call `load` directly, you just need to call [`from_request_parts`](https://docs.rs/axum-messages/latest/axum_messages/struct.Messages.html#method.from_request_parts); remember, `FromRequestParts` is already implemented for `Messages`, so once you have the request and state, you can invoke it yourself and cause the messages to be loaded.

Without seeing a minimal code example, it's admittedly difficult to help but I suspect you could implement your own extractor and barring that then you could do this in your handler with the request and state. I'm not sure exactly what you're doing with `map_response_with_state`, but if you need `Messages` there then again wrapping the extractor is an option (use `from_request_parts`).

Regardless of approach, my opinion is this is a job for the Axum API.

**Reply by marcaddeo** [reply]
> You don't need to call `load` directly, you just need to call [`from_request_parts`](https://docs.rs/axum-messages/latest/axum_messages/struct.Messages.html#method.from_request_parts); remember, `FromRequestParts` is already implemented for `Messages`, so once you have the request and state, you can invoke it yourself and cause the messages to be loaded.
> 
> Without seeing a minimal code example, it's admittedly difficult to help but I suspect you could implement your own extractor and barring that then you could do this in your handler with the request and state. I'm not sure exactly what you're doing with `map_response_with_state`, but if you need `Messages` there then again wrapping the extractor is an option (use `from_request_parts`).
> 
> Regardless of approach, my opinion is this is a job for the Axum API.

I don't have access to the `Request` in the `map_response_with_state` middleware, so I can't call `Messages::from_request_parts`. I'll put together an example.

**Reply by maxcountryman** [reply]
You probably want to use [`from_fn`](https://docs.rs/axum/latest/axum/middleware/fn.from_fn.html) in that case.

**Reply by marcaddeo** [reply]
Just attempted with `from_fn`, but don't think that will work because running `next.run(request)` consumes the request to produce the response and I need to check the response to determine if I need to consume the `Messages`.

[Here's a minimal example](https://github.com/marcaddeo/minimal-example), hopefully that helps.
